python/perf-kernels/flash-attention-seqlen-padded.py

```diff
@@ -0,0 +1,772 @@
+"""
+Fused Attention
+===============
+
+This is a Triton implementation of the Flash Attention v2 algorithm from Tri Dao (https://tridao.me/publications/flash2/flash2.pdf)
+Credits: OpenAI kernel team
+
+Extra Credits:
+- Original flash attention paper (https://arxiv.org/abs/2205.14135)
+- Rabe and Staats (https://arxiv.org/pdf/2112.05682v2.pdf)
+
+"""
+
+import pytest
+import torch
+
+
+import triton
+import triton.language as tl
+
+torch_dtype:tl.constexpr = torch.float16
+TORCH_HAS_FP8E5 = hasattr(torch, 'float8_e5m2fnuz')
+if TORCH_HAS_FP8E5:
+    torch_dtype:tl.constexpr = torch.float8_e5m2fnuz
+
+@triton.jit
+def _attn_fwd_inner(
+    acc, l_i, m_i, q,
+    K_block_ptr, V_block_ptr,
+    start_m,
+    BLOCK_M: tl.constexpr,
+    BLOCK_DMODEL: tl.constexpr,
+    BLOCK_N: tl.constexpr,
+    STAGE: tl.constexpr,
+    offs_m: tl.constexpr,
+    offs_n: tl.constexpr,
+    N_CTX,
+    pre_load_v: tl.constexpr,
+    padded_block: tl.constexpr,
+    total_tokens: tl.constexpr,
+):
+    # range of values handled by this stage
+    if STAGE == 1:
+        lo, hi = 0, start_m * BLOCK_M
+    elif STAGE == 2:
+        lo, hi = start_m * BLOCK_M, (start_m + 1) * BLOCK_M
+        lo = tl.multiple_of(lo, BLOCK_M)
+        K_block_ptr = tl.advance(K_block_ptr, (0, lo))
+        V_block_ptr = tl.advance(V_block_ptr, (lo, 0))
+    # N_CTX is the seqlen to the nearest block (round down).
+    # So here, we are computing the elements for that last irregular block.
+    # In the loop,  we will mask the elements of BLOCK_N that do not exist.
+    elif padded_block:
+        lo, hi = N_CTX, N_CTX + BLOCK_N
+        lo = tl.multiple_of(lo, BLOCK_M)
+        K_block_ptr = tl.advance(K_block_ptr, (0, lo))
+        V_block_ptr = tl.advance(V_block_ptr, (lo, 0))
```

Comment:
When padded_block is true, this is the second `_attn_fwd_inner` loop. At the end of the first `_attn_fwd_inner` loop, K/V_block_ptr should already be advanced at the correct location. 
Or are Triton function arguments passed by value?

Reply:
I assumed they were passed by value, because the causal=True case also advances block pointers in this way for STAGE=2:

https://github.com/ROCmSoftwarePlatform/triton/blob/non_pow_2_seqlen/python/perf-kernels/flash-attention-seqlen-padded.py#L50

And I assumed if I was wrong it'd segfault.